### PR TITLE
Add XML attributes for picker components jmix-framework/jmix#1787

### PR DESCRIPTION
**Pull request by KremnevDmitry**
Relate to: https://github.com/jmix-framework/jmix/issues/1787

### CODE REVIEW

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/DatePickerLoader.java [hunk 1]
```diff
@@ -55,6 +67,26 @@ public void loadComponent() {
         componentLoader().loadAllowedCharPattern(resultComponent, element, context);
     }
 
+    protected void loadLocalDate(Element element, String attributeName, Consumer<LocalDate> setter) {
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/DatePickerLoader.java:70]: loadDate

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/DatePickerLoader.java [hunk 2]
```diff
@@ -55,6 +67,26 @@ public void loadComponent() {
...
+        loadResourceString(element, attributeName, context.getMessageGroup())
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/DatePickerLoader.java:71]: regular string

File: jmix-flowui/flowui-kit/src/main/java/io/jmix/flowui/kit/meta/component/StudioComponents.java [hunk 3]
```diff
@@ -613,16 +613,20 @@ public interface StudioComponents {
...
                     @StudioProperty(xmlAttribute = "enabled", type = StudioPropertyType.BOOLEAN,
                             defaultValue = "true"),
                     @StudioProperty(xmlAttribute = "errorMessage", type = StudioPropertyType.LOCALIZED_STRING),
                     @StudioProperty(xmlAttribute = "height", type = StudioPropertyType.SIZE),
+                    @StudioProperty(xmlAttribute = "helperText", type = StudioPropertyType.LOCALIZED_STRING),
                     @StudioProperty(xmlAttribute = "id", type = StudioPropertyType.COMPONENT_ID),
                     @StudioProperty(xmlAttribute = "invalid", type = StudioPropertyType.BOOLEAN,
                             defaultValue = "false"),
                     @StudioProperty(xmlAttribute = "label", type = StudioPropertyType.LOCALIZED_STRING),
+                    @StudioProperty(xmlAttribute = "max", type = StudioPropertyType.LOCALIZED_STRING),
```
Comment (glebfox) [inline jmix-flowui/flowui-kit/src/main/java/io/jmix/flowui/kit/meta/component/StudioComponents.java:626]: regular string

File: jmix-flowui/flowui-kit/src/main/java/io/jmix/flowui/kit/meta/component/StudioComponents.java [hunk 4]
```diff
@@ -613,16 +613,20 @@ public interface StudioComponents {
...
                     @StudioProperty(xmlAttribute = "maxHeight", type = StudioPropertyType.SIZE),
                     @StudioProperty(xmlAttribute = "maxWidth", type = StudioPropertyType.SIZE),
+                    @StudioProperty(xmlAttribute = "min", type = StudioPropertyType.LOCALIZED_STRING),
```
Comment (glebfox) [inline jmix-flowui/flowui-kit/src/main/java/io/jmix/flowui/kit/meta/component/StudioComponents.java:629]: regular string

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/TimePickerLoader.java [hunk 5]
```diff
@@ -47,11 +55,31 @@ public void loadComponent() {
         componentLoader().loadRequired(resultComponent, element, context);
         componentLoader().loadTabIndex(resultComponent, element);
         componentLoader().loadThemeNames(resultComponent, element);
+        componentLoader().loadStep(resultComponent, element, context);
         componentLoader().loadValueAndElementAttributes(resultComponent, element);
         componentLoader().loadValidationAttributes(resultComponent, element, context);
         componentLoader().loadAllowedCharPattern(resultComponent, element, context);
     }
 
+    protected void loadLocalTime(Element element, String attributeName, Consumer<LocalTime> setter) {
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/TimePickerLoader.java:64]: loadTime

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/TimePickerLoader.java [hunk 6]
```diff
@@ -47,11 +55,31 @@ public void loadComponent() {
...
+        loadResourceString(element, attributeName, context.getMessageGroup())
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/TimePickerLoader.java:65]: regular string

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/DateTimePickerLoader.java [hunk 7]
```diff
@@ -48,10 +57,30 @@ public void loadComponent() {
         componentLoader().loadHelperText(resultComponent, element);
         componentLoader().loadSizeAttributes(resultComponent, element);
         componentLoader().loadRequired(resultComponent, element, context);
+        componentLoader().loadStep(resultComponent, element, context);
         componentLoader().loadValueAndElementAttributes(resultComponent, element);
         componentLoader().loadValidationAttributes(resultComponent, element, context);
     }
 
+    protected void loadLocalDateTime(Element element, String attributeName, Consumer<LocalDateTime> setter) {
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/loader/component/DateTimePickerLoader.java:65]: you know what to do...

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/support/ComponentLoaderSupport.java [hunk 8]
```diff
@@ -353,6 +358,51 @@ public void loadAllowedCharPattern(HasAllowedCharPattern component, Element elem
                 context.getMessageGroup(), component::setAllowedCharPattern);
     }
 
+    public void loadStep(com.vaadin.flow.component.Component component, Element element, Context context) {
+        loaderSupport.loadResourceString(element, "step", context.getMessageGroup())
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/support/ComponentLoaderSupport.java:362]: regulat string

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/support/ComponentLoaderSupport.java [hunk 9]
```diff
@@ -353,6 +358,51 @@ public void loadAllowedCharPattern(HasAllowedCharPattern component, Element elem
                 context.getMessageGroup(), component::setAllowedCharPattern);
     }
 
+    public void loadStep(com.vaadin.flow.component.Component component, Element element, Context context) {
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/support/ComponentLoaderSupport.java:361]: `void` -> `Duration`

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/support/ComponentLoaderSupport.java [hunk 10]
```diff
@@ -353,6 +358,51 @@ public void loadAllowedCharPattern(HasAllowedCharPattern component, Element elem
...
+                        step = Duration.ofHours(Long.parseLong(StringUtils.chop(stepString)));
+                    } else if (stepString.endsWith("m")) {
+                        step = Duration.ofMinutes(Long.parseLong(StringUtils.chop(stepString)));
+                    } else if (stepString.endsWith("s")) {
+                        step = Duration.ofSeconds(Long.parseLong(StringUtils.chop(stepString)));
+                    } else {
+                        step = Duration.ofMinutes(Long.parseLong(StringUtils.chop(stepString)));
+                    }
+
+                    if (component instanceof TimePicker timePicker) {
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/support/ComponentLoaderSupport.java:376]: move to component loaders, no more if-s

**Review by glebfox — CHANGES_REQUESTED** [review]

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/support/ComponentLoaderSupport.java [hunk 11]
```diff
@@ -358,9 +356,9 @@ public void loadAllowedCharPattern(HasAllowedCharPattern component, Element elem
                 context.getMessageGroup(), component::setAllowedCharPattern);
     }
 
-    public void loadStep(com.vaadin.flow.component.Component component, Element element, Context context) {
-        loaderSupport.loadResourceString(element, "step", context.getMessageGroup())
-                .ifPresent(stepString -> {
+    public Optional<Duration> loadStep(Element element) {
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/support/ComponentLoaderSupport.java:359]: Let's call it `loadDuration`. Who knows where else we can use this method.

**Review by glebfox — CHANGES_REQUESTED** [review]

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/support/ComponentLoaderSupport.java (hunk 11 above)
Comment (KremnevDmitry) [reply jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/support/ComponentLoaderSupport.java:359]: i.e. you want the attribute name to also be passed to the method?

Comment (glebfox) [reply jmix-flowui/flowui/src/main/java/io/jmix/flowui/xml/layout/support/ComponentLoaderSupport.java:359]: da

**Review by glebfox — APPROVED** [review]

**Pull request merged**